src/composed/message/types.rs

```diff
@@ -433,34 +438,57 @@ impl Message {
                     SignatureType::Text
                 };
 
-                let signature_config = SignatureConfig::new_v4(
-                    Default::default(),
-                    typ,
-                    algorithm,
-                    hash_algorithm,
-                    hashed_subpackets,
-                    unhashed_subpackets,
-                );
-                (typ, signature_config.sign(key, key_pw, l.data())?)
+                let mut config = match sig_version {
+                    SignatureVersion::V4 => SignatureConfig::v4(typ, algorithm, hash_algorithm),
+                    SignatureVersion::V6 => {
+                        SignatureConfig::v6_with_rng(rng, typ, algorithm, hash_algorithm)?
+                    }
+                    _ => bail!("unsupported signature version {:?}", sig_version),
+                };
+                config.hashed_subpackets = hashed_subpackets;
+                config.unhashed_subpackets = unhashed_subpackets;
+
+                (typ, config.sign(key, key_pw, l.data())?)
             }
             _ => {
                 let typ = SignatureType::Binary;
-                let signature_config = SignatureConfig::new_v4(
-                    Default::default(),
-                    typ,
-                    algorithm,
-                    hash_algorithm,
-                    hashed_subpackets,
-                    unhashed_subpackets,
-                );
+
+                let mut config = match sig_version {
+                    SignatureVersion::V4 => SignatureConfig::v4(typ, algorithm, hash_algorithm),
+                    SignatureVersion::V6 => {
+                        SignatureConfig::v6_with_rng(rng, typ, algorithm, hash_algorithm)?
+                    }
+                    _ => bail!("unsupported signature version {:?}", sig_version),
+                };
+                config.hashed_subpackets = hashed_subpackets;
+                config.unhashed_subpackets = unhashed_subpackets;
 
                 let data = self.to_bytes()?;
-                let signature = signature_config.sign(key, key_pw, &data[..])?;
+                let signature = config.sign(key, key_pw, &data[..])?;
 
                 (typ, signature)
             }
         };
-        let ops = OnePassSignature::from_details(typ, hash_algorithm, algorithm, key_id);
+
+        let ops = match sig_version {
+            SignatureVersion::V4 => {
+                OnePassSignature::from_details_v3(typ, hash_algorithm, algorithm, key_id)
```

Comment:
given that the others are called `v4()` `v6()` etc, should we call this just `v3()`?

Reply:
good catch. done.